x/evm/keeper/abci.go

```diff
@@ -17,17 +17,30 @@ package keeper
 
 import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	ethermint "github.com/evmos/ethermint/types"
 )
 
 // BeginBlock sets the sdk Context and EIP155 chain id to the Keeper.
 func (k *Keeper) BeginBlock(ctx sdk.Context) error {
 	k.WithChainID(ctx)
 
 	// cache parameters that's common for the whole block.
-	if _, err := k.EVMBlockConfig(ctx, k.ChainID()); err != nil {
+	cfg, err := k.EVMBlockConfig(ctx, k.ChainID())
+	if err != nil {
 		return err
 	}
-
+	k.SetHeaderHash(ctx)
+	headerHashNum, err := ethermint.SafeInt64(cfg.Params.GetHeaderHashNum())
+	if err != nil {
+		panic(err)
+	}
+	for i := ctx.BlockHeight() - headerHashNum; i >= 0; i-- {
```

Comment:
this is too heavy, just delete the `ctx.BlockHeight() - headerHashNum` if it's positive.